src/amberscript/parser_viewport_test.cc

```diff
@@ -0,0 +1,340 @@
+SHADER fragment my_fragment GLSL
+# GLSL Shader
+END
+BUFFER my_fb FORMAT R32G32B32A32_SFLOAT
+BUFFER my_ds FORMAT D32_SFLOAT_S8_UINT
+
+PIPELINE graphics my_pipeline
+  ATTACH my_shader
+  ATTACH my_fragment
+  BIND BUFFER my_fb AS color LOCATION 0
+  BIND BUFFER my_ds AS depth_stencil
+
+  VIEWPORT 12.2 9.7 SIZE 0.5 106.1 MAX_DEPTH 0.456
+END)";
+
+  Parser parser;
+  Result r = parser.Parse(in);
+  ASSERT_TRUE(r.IsSuccess()) << r.Error();
+
+  auto script = parser.GetScript();
+  const auto& pipelines = script->GetPipelines();
+  ASSERT_EQ(1U, pipelines.size());
+
+  auto* pipeline = pipelines[0].get();
+  ASSERT_TRUE(pipeline->GetPipelineData()->HasViewportData());
+  ASSERT_FLOAT_EQ(12.2f, pipeline->GetPipelineData()->GetViewportX());
+  ASSERT_FLOAT_EQ(9.7f, pipeline->GetPipelineData()->GetViewportY());
+  ASSERT_FLOAT_EQ(0.5f, pipeline->GetPipelineData()->GetViewportW());
+  ASSERT_FLOAT_EQ(106.1f, pipeline->GetPipelineData()->GetViewportH());
+  ASSERT_FLOAT_EQ(0.0f, pipeline->GetPipelineData()->GetViewportMinDepth());
+  ASSERT_FLOAT_EQ(0.456f, pipeline->GetPipelineData()->GetViewportMaxDepth());
+}
+
+TEST_F(AmberScriptParserTest, ViewportAllValues) {
+  std::string in = R"(
+SHADER vertex my_shader PASSTHROUGH
+SHADER fragment my_fragment GLSL
+# GLSL Shader
+END
+BUFFER my_fb FORMAT R32G32B32A32_SFLOAT
+BUFFER my_ds FORMAT D32_SFLOAT_S8_UINT
+
+PIPELINE graphics my_pipeline
+  ATTACH my_shader
+  ATTACH my_fragment
+  BIND BUFFER my_fb AS color LOCATION 0
+  BIND BUFFER my_ds AS depth_stencil
+
+  VIEWPORT -0.6 5.2 SIZE 13.8 9.4 MIN_DEPTH 0.5 MAX_DEPTH 0.6
+END)";
+
+  Parser parser;
+  Result r = parser.Parse(in);
+  ASSERT_TRUE(r.IsSuccess()) << r.Error();
+
+  auto script = parser.GetScript();
+  const auto& pipelines = script->GetPipelines();
+  ASSERT_EQ(1U, pipelines.size());
+
+  auto* pipeline = pipelines[0].get();
+  ASSERT_TRUE(pipeline->GetPipelineData()->HasViewportData());
+  ASSERT_FLOAT_EQ(-0.6f, pipeline->GetPipelineData()->GetViewportX());
+  ASSERT_FLOAT_EQ(5.2f, pipeline->GetPipelineData()->GetViewportY());
+  ASSERT_FLOAT_EQ(13.8f, pipeline->GetPipelineData()->GetViewportW());
+  ASSERT_FLOAT_EQ(9.4f, pipeline->GetPipelineData()->GetViewportH());
+  ASSERT_FLOAT_EQ(0.5f, pipeline->GetPipelineData()->GetViewportMinDepth());
+  ASSERT_FLOAT_EQ(0.6f, pipeline->GetPipelineData()->GetViewportMaxDepth());
+}
+
+TEST_F(AmberScriptParserTest, ViewportInvalidIntegers) {
+  std::string in = R"(
+SHADER vertex my_shader PASSTHROUGH
+SHADER fragment my_fragment GLSL
+# GLSL Shader
+END
+BUFFER my_fb FORMAT R32G32B32A32_SFLOAT
+BUFFER my_ds FORMAT D32_SFLOAT_S8_UINT
+
+PIPELINE graphics my_pipeline
+  ATTACH my_shader
+  ATTACH my_fragment
+  BIND BUFFER my_fb AS color LOCATION 0
+  BIND BUFFER my_ds AS depth_stencil
+
+  VIEWPORT 0 0 SIZE 15 20
+END)";
+
+  Parser parser;
+  Result r = parser.Parse(in);
+  ASSERT_FALSE(r.IsSuccess()) << r.Error();
+  EXPECT_EQ("15: invalid offset for VIEWPORT command", r.Error());
```

Comment:
Could we also allow integers and just convert them to floats? If not, at least the error message should be more specific.

Reply:
I'm fine with that, although I don't think it wasn't done like that in whatever command I copied this from. They have to be floats in Vulkan. Is there a general policy on how these commands work in amber?

Reply:
Actually, I found the places in the existing code that were doing this and have updated this change to allow integers as well.